File: PHP/One/Transforms/InferTypesPlugins/ResolveFuncCalls.php
<?php

namespace One\Transforms\InferTypesPlugins\ResolveFuncCalls;

use One\Transforms\InferTypesPlugins\Helpers\InferTypesPlugin\InferTypesPlugin;
use One\Ast\Expressions\Expression;
use One\Ast\Expressions\UnresolvedCallExpression;
use One\Ast\Expressions\GlobalFunctionCallExpression;
use One\Ast\Expressions\LambdaCallExpression;
use One\Ast\References\GlobalFunctionReference;
use One\Ast\AstTypes\LambdaType;

class ResolveFuncCalls extends InferTypesPlugin {
    function __construct() {
        parent::__construct("ResolveFuncCalls");
        
    }
    
    function canTransform($expr) {
        return $expr instanceof UnresolvedCallExpression;
    }
    
    function transform($expr) {
        $callExpr = $expr;
        if ($callExpr->func instanceof GlobalFunctionReference) {
            $newExpr = new GlobalFunctionCallExpression($callExpr->func->decl, $callExpr->args);
            $callExpr->args = array_map(function ($arg) { return $this->main->runPluginsOn($arg); }, $callExpr->args);
            $newExpr->setActualType($callExpr->func->decl->returns);
            return $newExpr;
        }
        else {
            $this->main->processExpression($expr);
            if ($callExpr->func->actualType instanceof LambdaType) {
                $newExpr = new LambdaCallExpression($callExpr->func, $callExpr->args);
                $newExpr->setActualType($callExpr->func->actualType->returnType);
                return $newExpr;
            }
            else
                return $expr;
        }
    }
}

File: CSharp/One/Transforms/InferTypesPlugins/ResolveFuncCalls.cs
using One.Ast;
using One.Transforms.InferTypesPlugins.Helpers;

namespace One.Transforms.InferTypesPlugins
{
    public class ResolveFuncCalls : InferTypesPlugin {
        public ResolveFuncCalls(): base("ResolveFuncCalls")
        {
            
        }
        
        public override bool canTransform(Expression expr)
        {
            return expr is UnresolvedCallExpression;
        }
        
        public override Expression transform(Expression expr)
        {
            var callExpr = ((UnresolvedCallExpression)expr);
            if (callExpr.func is GlobalFunctionReference globFunctRef) {
                var newExpr = new GlobalFunctionCallExpression(globFunctRef.decl, callExpr.args);
                callExpr.args = callExpr.args.map(arg => this.main.runPluginsOn(arg));
                newExpr.setActualType(globFunctRef.decl.returns);
                return newExpr;
            }
            else {
                this.main.processExpression(expr);
                if (callExpr.func.actualType is LambdaType lambdType) {
                    var newExpr = new LambdaCallExpression(callExpr.func, callExpr.args);
                    newExpr.setActualType(lambdType.returnType);
                    return newExpr;
                }
                else
                    return expr;
            }
        }
    }
}